docs/design.md

```diff
@@ -0,0 +1,206 @@
+
+This document explains the purpose of the sops-compliance-checker and also
+contains all the details about design decisions we took. It assumes familarity
+with [SOPS][sops] and the concepts that it uses.
+
+## Problem Statement
+
+SOPS supports encrypting files using one or more trust anchors. Especially in
+enterprise contexts, it is important to enforce some rules around the trust
+anchors that are in use. Some questions that are looking for an answer could
+be:
+
+- How can we ensure that any given SOPS file is encrypted using a set of
+  required trust anchors?
+- How can we detect if non-approved trust anchors are used?
+- How can we prevent the usage of certain trust anchors in the presence of
+  others (mutual exclusivity)?
+
+To answer these questions for any given SOPS file, the
+`sops-compliance-checker` needs to be flexible enough to process a set of
+user-defined rules which indicate success or failure.
+
+## Functional Requirements
+
+The high-level functionality of the compliance checker can be summarized like
+this:
+
+- Recursively scan a directory tree (such as a Git repository) for SOPS
+  encrypted files. It should support all file formats supported by SOPS itself.
+- Extract trust anchors from SOPS encrypted files in order to match them
+  against a set of user-defined rules.
+- Report check results back to the user.
+
+The rule engine needs to support the following functionality:
+
+- Matching of a trust anchor ("match")
+  - **Example**: The trust anchor is an AWS KMS Key with the ARN
+    `arn:aws:kms:eu-west-1:123456789012:alias/my-team`.
+- Matching of all rules ("and" / "all of")
+  - **Example**: The SOPS file must be encrypted with the AGE disaster recovery
+    key as well as the keys of the owner of the file and the deploy key for the
+    CI/CD to be able to decrypt it.
+- Matching of any rule ("or" / "any of")
+  - **Example**: The SOPS file must be encrypted using the key of one or more
+    teams responsible for the software component.
+- Matching of exactly one rule ("xor" / "one of")
+  - **Example**: The trust anchors can only contain the deploy key for
+    production **or** staging, but not both, to prevent secret sharing between
+    environments.
+- Inversion of match behaviour ("not")
+  - **Example**: There's an explicit key that should not be part of the trust
+    anchors.
+- Reject excess trust anchors (not matched by any rule):
+  - **Example**: Developers should not be allowed to use additional encryption
+    keys apart from the keys managed by the company within any given SOPS file.
+    This protects against employees still retaining access to secrets after
+    they left the company.
+
+It must be possible to nest these rule arbitrarily deep to allow building
+complex match pattern with different dependencies between each other.
+
+## Non-functional Requirements
+
+- The compliance checker should be packaged as a container image for easy
+  integration into CI/CD pipelines.
+
+## Data Model
```

Comment:
FYI: https://github.com/CacheControl/json-rules-engine/tree/master

This is also used at [`@backstage-community/plugin-tech-insights-backend-module-jsonfc`](https://www.npmjs.com/package/@backstage-community/plugin-tech-insights-backend-module-jsonfc)

Reply:
another option could be [Rego by OPA](https://www.openpolicyagent.org/docs/latest/policy-language/)

Reply:
We could think about adding a section around existing rules engines and compare them with whatever we settle for at the end.